pkg/prove/proof.go

```diff
@@ -203,3 +206,105 @@ func splitIntoRows(squareSize uint64, s []shares.Share) [][]shares.Share {
 	}
 	return rows
 }
+
+// SharesInclusion generates an nmt inclusion proof for a set of shares to the data root.
+// expects the shares range to be pre-validated.
+// Note: only supports inclusion proofs for shares belonging to the same namespace.
+func SharesInclusion(
+	codec rsmt2d.Codec,
+	allRawShares []shares.Share,
+	squareSize uint64,
+	namespaceID namespace.ID,
+	startShare uint64,
+	endShare uint64,
+) (types.SharesProof, error) {
+	startRow := startShare / squareSize
+	endRow := endShare / squareSize
+	startLeaf := startShare % squareSize
+	endLeaf := endShare % squareSize
+
+	eds, err := rsmt2d.ComputeExtendedDataSquare(
+		shares.ToBytes(allRawShares),
+		appconsts.DefaultCodec(),
+		wrapper.NewConstructor(squareSize),
+	)
+	if err != nil {
+		return types.SharesProof{}, err
+	}
+
+	// create the binary merkle inclusion proof, for all the square rows, to the data root
+	rootHash, allProofs := merkle.ProofsFromByteSlices(append(eds.RowRoots(), eds.ColRoots()...))
+	rowsProofs := make([]*merkle.Proof, endRow-startRow+1)
+	rowsRoots := make([]tmbytes.HexBytes, endRow-startRow+1)
+	for i := startRow; i <= endRow; i++ {
+		rowsProofs[i-startRow] = allProofs[i]
+		rowsRoots[i-startRow] = eds.RowRoots()[i]
+		// verifying that the proofs are correct
+		err := rowsProofs[i-startRow].Verify(rootHash, rowsRoots[i-startRow])
```

Comment:
[note for reviewers]
I am verifying that the generated proofs are correct at this level, before continuing the execution. Which better, making sure that the returned proof is correct before returning to the user? Or let the user validate his proofs himself?

Reply:
[96c4257](https://github.com/celestiaorg/celestia-app/pull/1042/commits/96c42577e91e72d68fa73ae5a67e1213fc5b602a)